src/goto-instrument/code_contracts.cpp

```diff
@@ -244,6 +245,106 @@ void code_contractst::add_quantified_variable(
   }
 }
 
+const optionalt<exprt> code_contractst::replace_old_parameter(
+  exprt expr,
```

Comment:
Could you make this `exprt &expr` to avoid a copy. I see that it can't be `const exprt&` since you're modifying it.

Also, this pattern looks bad:
- you pass `expression` to this function, which could be modified within the function
- then you return an `optional<exprt>` from this function
- you replace `expression` if the `optional<exprt>` isn't nil

Why not just pass `exprt &expr` as first argument here and return `void`?
This call that you do later:

```c
  // Find and replace "old" expression in the "expression" variable
  const auto expr_w_replacement = replace_old_parameter(
    expression, parameter2history, location, function, mode, history);
  if(expr_w_replacement.has_value())
  {
    expression = to_code(*expr_w_replacement);
  }
```

could just be one line: `replace_old_parameter(expression, parameter2history, location, function, mode, history);` -- `expression` would be passed by reference, so would be modified within `replace_old_parameter` as needed.